File: frontend/src/features/suiji/editor-model.ts
import {
  SUIJI_LIMITS,
  type RecordResponse,
  type SuijiRecord,
  type UploadedAttachment,
} from "@runweave/shared/suiji";
import { SuijiClient, SuijiHttpError } from "../../services/suiji";
import { SuijiDraftStore, type SuijiDraft } from "./drafts";

export class SuijiEditorModel {
  private listeners = new Set<() => void>();
  private queue: Promise<unknown> = Promise.resolve();
  private writes = 0;
  private active = true;
  state: {
    draft: SuijiDraft;
    busy: boolean;
    localSaving: boolean;
    message: string;
    localError: boolean;
    conflict: boolean;
    latest?: SuijiRecord;
    saved?: SuijiRecord;
    discarded?: boolean;
  };
  constructor(
    draft: SuijiDraft,
    private client: SuijiClient,
    private store: SuijiDraftStore,
  ) {
    this.state = {
      draft,
      busy: false,
      localSaving: false,
      message: draft.frozen ? "保存结果待确认，请手动重试；输入暂时锁定" : "",
      localError: false,
      conflict: false,
    };
  }
  subscribe = (callback: () => void) => {
    this.listeners.add(callback);
    return () => {
      this.listeners.delete(callback);
    };
  };
  snapshot = () => this.state;
  dispose() {
    this.active = false;
    this.listeners.clear();
  }
  settled() {
    return this.queue;
  }
  initialize() {
    return this.persist();
  }
  async discard() {
    if (this.state.busy || this.state.draft.frozen) return;
    this.patch({ busy: true });
    try { await this.queue; this.check(); await this.store.remove("draft:" + this.state.draft.id); this.patch({ discarded: true }); }
    catch (error) { this.patch({ message: error instanceof Error ? error.message : "放弃草稿失败" }); }
    finally { this.patch({ busy: false }); }
  }
  private check() {
    if (!this.active || !this.client.active)
      throw new DOMException("编辑会话已结束", "AbortError");
  }
  private patch(update: Partial<typeof this.state>) {
    this.state = { ...this.state, ...update };
    this.listeners.forEach((f) => f());
  }
  private persist(draft = this.state.draft) {
    const snapshot = structuredClone(draft);
    ++this.writes;
    this.patch({ localSaving: true });
    const pending = this.queue.then(() =>
      this.store.set("draft:" + snapshot.id, snapshot),
    );
    this.queue = pending.catch(() => undefined);
    return pending
      .then(() => this.patch({ localError: false }))
      .catch((error: unknown) => {
        this.patch({
          message: error instanceof Error ? error.message : "本机草稿未保存",
          localError: true,
        });
        throw error;
      })
      .finally(() => {
        --this.writes;
        this.patch({ localSaving: this.writes > 0 });
      });
  }
  edit(
    update: Partial<Pick<SuijiDraft, "kind" | "body" | "existing" | "files">>,
  ) {
    if (this.state.busy || this.state.draft.frozen) return;
    const draft = { ...this.state.draft, ...update };
    this.patch({ draft });
    void this.persist(draft).catch(() => undefined);
  }
  async addFile(file: File) {
    try {
      this.check();
      if (file.size > SUIJI_LIMITS.attachmentBytes)
        throw new Error("每个附件最多 5 MiB");
      const image = ["image/jpeg", "image/png"].includes(file.type);
      if (!image && !/\.md$/i.test(file.name))
        throw new Error("仅支持 JPEG、PNG 和 Markdown");
      if (!image)
        new TextDecoder("utf-8", { fatal: true }).decode(
          await file.arrayBuffer(),
        );
      this.check();
      const draft = this.state.draft,
        kind = image ? "image" : "markdown";
      if (
        draft.existing.some((a) => a.kind === kind) ||
        draft.files.some(
          (a) =>
            (a.file.type.startsWith("image/") ? "image" : "markdown") === kind,
        )
      )
        throw new Error("每条最多一张图片和一个 Markdown，请先移除原附件");
      if (!image) file = new File([file], file.name, { type: "text/markdown", lastModified: file.lastModified });
      this.edit({
        files: [
          ...draft.files,
          { id: crypto.randomUUID(), file, key: crypto.randomUUID() },
        ],
      });
    } catch (error) {
      this.patch({
        message: error instanceof Error ? error.message : "附件读取失败",
      });
    }
  }
  async save() {
    if (this.state.busy || this.state.saved) return;
    this.patch({ busy: true, message: "" });
    try {
      this.check();
      let draft = this.state.draft;
      if ([...draft.body].length > SUIJI_LIMITS.bodyScalars)
        throw new Error("正文最多 20,000 个字符");
      if (!draft.body.trim() && !draft.existing.length && !draft.files.length)
        throw new Error("请输入正文或添加附件");
      await this.persist(draft);
      this.check();
      if (!draft.pending) {
        draft = { ...draft, frozen: true };
        this.patch({ draft });
        await this.persist(draft);
        for (const [index, item] of draft.files.entries()) {
          if (item.uploaded) continue;
          this.check();
          const form = new FormData();
          form.append("file", item.file, item.file.name);
          const result = await this.client.request<{
            attachment: UploadedAttachment;
          }>("/api/suiji/v1/uploads", "POST", form, item.key);
          this.check();
          draft = {
            ...draft,
            files: draft.files.map((file, i) =>
              i === index ? { ...file, uploaded: result.attachment } : file,
            ),
          };
          this.patch({ draft });
          await this.persist(draft);
        }
        draft = {
          ...draft,
          pending: {
            path:
              "/api/suiji/v1/records" +
              (draft.id === "new" ? "" : "/" + draft.id),
            method: draft.id === "new" ? "POST" : "PATCH",
            key: crypto.randomUUID(),
            data: {
              kind: draft.kind,
              ...(draft.id === "new" ? {} : { expectedVersion: draft.version }),
              body: draft.body,
              attachmentIds: [
                ...draft.existing.map((a) => a.id),
                ...draft.files.map((a) => a.uploaded!.id),
              ],
            },
          },
        };
        this.patch({ draft });
        await this.persist(draft);
      }
      this.check();
      const pending = draft.pending!;
      const result = await this.client.request<RecordResponse>(
        pending.path,
        pending.method,
        pending.data,
        pending.key,
      );
      this.check();
      await this.store.remove("draft:" + draft.id);
      this.patch({ saved: result.record });
    } catch (error) {
      if (!this.active) return;
      if (error instanceof SuijiHttpError && !error.uncertain) {
        const conflict = error.detail.error.code === "VERSION_CONFLICT";
        const draft = {
          ...this.state.draft,
          frozen: false,
          pending: undefined,
        };
        this.patch({ draft, conflict, message: error.message });
        await this.persist(draft).catch(() => undefined);
      } else
        this.patch({
          message:
            (this.state.draft.frozen
              ? "保存结果待确认，手动重试会沿用同一次请求。"
              : "") + (error instanceof Error ? error.message : "保存失败"),
        });
    } finally {
      this.patch({ busy: false });
    }
  }
  async loadLatest() {
    if (this.state.busy || this.state.draft.id === "new") return;
    try {
      const { record } = await this.client.request<RecordResponse>(
        "/api/suiji/v1/records/" + this.state.draft.id,
      );
      this.check();
      this.patch({ latest: record });
    } catch (error) {
      this.patch({
        message: error instanceof Error ? error.message : "读取最新版本失败",
      });
    }
  }
  acceptLatestVersion() {
    const latest = this.state.latest;
    if (!latest || this.state.busy || this.state.draft.frozen) return;
    // Keep local body and kind: use the current remote attachments so concurrent file edits are not silently undone.
    const draft = {
      ...this.state.draft,
      version: latest.version,
      existing: latest.attachments,
      files: [],
    };
    if (this.state.draft.files.length) {
      this.patch({
        message: "本机还有新增附件，请先保存其文件并移除，再采用最新附件版本",
      });
      return;
    }
    this.patch({
      draft,
      latest: undefined,
      conflict: false,
      message: "已采用最新版本号与附件，本机正文和类型保留；请比较后手动保存",
    });
    void this.persist(draft).catch(() => undefined);
  }
}

File: packages/suiji-ios/Sources/SuijiIOS/State/EditorModel.swift
import Foundation
import SwiftUI
@MainActor final class EditorModel: ObservableObject, Identifiable {
  let id = UUID()
  @Published var draft: Draft
  @Published var busy = false
  @Published var message = ""
  @Published var latest: SuijiRecord?
  @Published var confirmed = false
  let client: APIClient
  let store: DraftStore
  let limits: Limits
  private var active = true
  init(draft: Draft, client: APIClient, store: DraftStore, limits: Limits) {
    self.draft = draft; self.client = client; self.store = store; self.limits = limits
  }
  var canReopen: Bool { active && !confirmed }
  var editable: Bool { !busy && !draft.frozen && !confirmed && active }
  func cancel() { active = false }
  func prepareForCapture(kind: RecordKind, body: String) async {
    guard draft.recordID == nil else { return }
    if editable, !draft.conflict, draft.pending == nil, draft.body.isEmpty, draft.existing.isEmpty, draft.local.isEmpty {
      guard draft.kind != kind || !body.isEmpty else { return }
      draft.kind = kind; draft.body = body; await persist()
    } else if !body.isEmpty {
      message = "已恢复原有草稿，新内容未覆盖它。请先处理当前草稿，再另存。"
    }
  }
  func persist() async {
    draft.revision += 1; let snapshot = draft
    do { try await store.save(snapshot); if active && draft.revision == snapshot.revision { message = "草稿已保存在本机" } }
    catch { if active { message = "草稿写入本机失败，请保留页面：" + error.localizedDescription } }
  }
  func add(data: Data, name: String, mime: String, kind: String) async {
    guard editable else { return }
    let count = draft.existing.filter { $0.kind == kind }.count + draft.local.filter { $0.kind == kind }.count
    let max = kind == "image" ? limits.imagesPerRecord : limits.markdownPerRecord
    guard count < max else { message = "此类附件已达上限，请先明确移除已有附件"; return }
    do {
      let item = try await store.importFile(data: data, fileName: name, mimeType: mime, kind: kind, limit: limits.attachmentBytes)
      guard active, editable else { return }; draft.local.append(item); await persist()
    } catch { message = error.localizedDescription }
  }
  func removeLocal(_ item: LocalAttachment) async {
    guard editable else { return }
    draft.local.removeAll { $0.id == item.id }; draft.revision += 1
    do { try await store.save(draft); try await store.removeLocal(item); message = "草稿已保存在本机" }
    catch { message = "移除附件后写盘失败：" + error.localizedDescription }
  }
  func save() async {
    guard !busy, active, !confirmed else { return }
    guard !draft.conflict else { message = "请先查看并处理版本冲突"; return }
    busy = true; defer { busy = false }
    do {
      guard draft.body.unicodeScalars.count <= limits.bodyScalars, !draft.body.contains("\0") else { throw MessageError(message: "正文超出限额或包含无效字符") }
      guard !draft.body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || !draft.local.isEmpty || !draft.existing.isEmpty else { throw MessageError(message: "请输入正文或添加附件") }
      draft.frozen = true; draft.revision += 1; try await store.save(draft)
      for index in draft.local.indices where draft.local[index].uploaded == nil {
        try checkActive()
        let item = draft.local[index], bytes = try await store.data(item), boundary = UUID().uuidString
        var multipart = Data("--\(boundary)\r\nContent-Disposition: form-data; name=\"file\"; filename=\"\(item.fileName.replacingOccurrences(of: "\"", with: "_").replacingOccurrences(of: "\r", with: "_").replacingOccurrences(of: "\n", with: "_"))\"\r\nContent-Type: \(item.mimeType)\r\n\r\n".utf8)
        multipart.append(bytes); multipart.append(Data("\r\n--\(boundary)--\r\n".utf8))
        let result = try await client.request(UploadResponse.self, path: "api/suiji/v1/uploads", method: "POST", data: multipart, key: item.uploadKey, contentType: "multipart/form-data; boundary=\(boundary)")
        try checkActive(); draft.local[index].uploaded = result.attachment; draft.revision += 1; try await store.save(draft)
      }
      if draft.pending == nil {
        let ids = draft.existing.map(\.id) + draft.local.compactMap { $0.uploaded?.id }
        var payload: [String: Any] = ["kind": draft.kind.rawValue, "body": draft.body, "attachmentIds": ids]
        if let version = draft.expectedVersion { payload["expectedVersion"] = version }
        draft.pending = PendingOperation(path: "api/suiji/v1/records" + (draft.recordID.map { "/" + $0 } ?? ""), method: draft.recordID == nil ? "POST" : "PATCH", payload: try JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys]))
        draft.revision += 1; try await store.save(draft)
      }
      try checkActive(); guard let operation = draft.pending else { return }
      let result = try await client.request(RecordResponse.self, path: operation.path, method: operation.method, data: operation.payload, key: operation.key)
      try checkActive()
      guard result.record.body == draft.body, result.record.kind == draft.kind,
        draft.recordID == nil || result.record.id == draft.recordID,
        UUID(uuidString: result.record.id) != nil, result.record.version >= (draft.expectedVersion ?? 1),
        result.record.attachments.map(\.id) == draft.existing.map(\.id) + draft.local.compactMap({ $0.uploaded?.id }) else { throw MessageError(message: "响应与保存内容不一致，请重试确认") }
      try await store.remove(draft); confirmed = true; message = "已保存"
    } catch is CancellationError { return }
    catch let error as APIError {
      guard active else { return }
      if error.error.code == "VERSION_CONFLICT" {
        draft.frozen = false; draft.pending = nil; draft.conflict = true
        message = "版本冲突，草稿保留在本机。请查看最新内容后决定。"
      } else if !error.uncertain {
        draft.frozen = false; draft.pending = nil; message = error.localizedDescription
      } else { message = "保存结果待确认，请手动重试确认。" }
      draft.revision += 1
      do { try await store.save(draft) } catch { message += " 本机状态写入失败，请保留页面。" }
    } catch { if active { message = draft.frozen ? "保存结果待确认：\(error.localizedDescription)" : error.localizedDescription } }
  }
  private func checkActive() throws { if !active { throw CancellationError() } }
  func compare() async {
    guard let id = draft.recordID, active else { return }
    do { let result = try await client.request(RecordResponse.self, path: "api/suiji/v1/records/" + id); if active { latest = result.record } }
    catch { if active { message = error.localizedDescription } }
  }
  func rebase() async {
    guard let latest, active else { return }
    draft.expectedVersion = latest.version; draft.existing = latest.attachments
    // A retained local upload already bound in the remote result must not be added twice.
    draft.local.removeAll { item in latest.attachments.contains { $0.id == item.uploaded?.id } }
    draft.conflict = false; draft.pending = nil; draft.frozen = false; self.latest = nil; await persist()
  }
  func discard() async -> Bool {
    guard editable else { return false }
    do { try await store.remove(draft); active = false; return true } catch { message = error.localizedDescription; return false }
  }
}
